test/tools/yulopti.cpp

```diff
@@ -171,28 +171,152 @@ class YulOpti
 		}
 	}
 
-	void disambiguate()
+	void applyFunctionToObjectAndSubobjects(Object& _object, function<void(Object&)> _function)
 	{
-		*m_ast = std::get<yul::Block>(Disambiguator(m_dialect, *m_analysisInfo)(*m_ast));
-		m_analysisInfo.reset();
-		m_nameDispenser.reset(*m_ast);
+		for (auto const& subObjectNode: _object.subObjects)
+		{
+			auto subObject = dynamic_pointer_cast<Object>(subObjectNode);
+
+			if (subObject != nullptr)
+				applyFunctionToObjectAndSubobjects(*subObject, _function);
+		}
+
+		_function(_object);
+	}
+
+	void runCodeAnalyzer(ErrorReporter& _errorReporter)
+	{
+		applyFunctionToObjectAndSubobjects(
+			*m_object,
+			[&](Object& _object)
+			{
+				_object.analysisInfo = make_shared<yul::AsmAnalysisInfo>();
+
+				AsmAnalyzer analyzer(
+					*_object.analysisInfo,
+					_errorReporter,
+					m_dialect,
+					{},
+					_object.qualifiedDataNames()
+				);
+
+				bool success = analyzer.analyze(*_object.code);
+				yulAssert(success && !_errorReporter.hasErrors(), "Invalid assembly/yul code.");
```

Comment:
Before this PR an analysis error resulted in the errors being printed with `printErrors()`. I think we should keep that (unless I'm missing something and this behavior is somehow preserved?).

It would also throw `std::runtime_error`, which is semantically different from an assertion, even though currently both are being handled the same way in `catch (...)`. We should really refactor this some day to throw a more specific exception class on validation/runtime errors and handle them differently from failed assertions.